readme
                                                                     
                                                                     
                                                                     
                                             


TO COMPILE THE CODE

	make


To view Manfile:
   $more man_hfd_output

TO RUN:

	$ hfd -h
	$ hfd  <input file>
	$ hfd -u <input file>

For Latest Version:
	https://github.com/erantanen/hfd

FILES USED:
	hfd.c
 	hfd_man
	makefile
	man_hfd_output




changes.txt

// A place to keep track of dev or changes.



2012 nov 15	:began getopt change over with initial variable.


fread-test.c
#include <stdio.h>


void main()
{

 int i;
 int b =16;
 int numberofElements = 1;
 int elementGroup = 0;

 unsigned char  n[b];
 FILE *filehandle_in;

 filehandle_in = fopen("a.out", "r");
 
 while( numberofElements != 0) {

  numberofElements = fread(n, sizeof(char), 16, filehandle_in);
    
 /* begin internal print block */

   printf("\n %06x:", elementGroup);

   for(i = 0; i < b ; i+=2){
	   printf(" %02X%02X ", n[i], n[i+1]);

    } 

    printf("\t");

    for(i = 0; i < b ; i++){

     /*  usable ascii symbols are 33 to 126 */
      if ( n[i] < 32 || n[i] > 126 ){
          printf(".");
        }
     else 
       {
         printf("%c", (char)n[i]);
       }

    }

   elementGroup = elementGroup + 16;
 }

/* end of internal print block */

    printf("\n\n");



fclose(filehandle_in);

 
}


hfd.s
	.file	"hfd.c"
	.globl	version
	.data
	.align 32
	.type	version, @object
	.size	version, 40
version:
	.string	"hfd v001.1 20sep12 by Ed Rantanen"
	.zero	6
	.section	.rodata
	.align 4
.LC0:
	.string	"\n 1 In correct usage,\n ' hfd -help '"
	.align 4
.LC1:
	.string	"Try for a better understanding of how to use hfd.\n"
.LC2:
	.string	" %02x%02x "
.LC3:
	.string	"\n %06x:"
.LC4:
	.string	"-u"
.LC5:
	.string	" %02X%02X "
.LC6:
	.string	"r"
	.align 4
.LC7:
	.string	"\n 2 In correct usage,\n ' hfd -help '"
.LC8:
	.string	"-ps"
.LC9:
	.string	"%02x%02x"
.LC10:
	.string	"\n"
.LC11:
	.string	"-v"
.LC12:
	.string	"\n%s \n\n"
.LC13:
	.string	"-c"
.LC14:
	.string	"-h"
.LC15:
	.string	"man ./hfd_man"
.LC16:
	.string	"Unable to locate help file"
	.text
	.globl	main
	.type	main, @function
main:
.LFB0:
	.cfi_startproc
	leal	4(%esp), %ecx
	.cfi_def_cfa 1, 0
	andl	$-16, %esp
	pushl	-4(%ecx)
	pushl	%ebp
	movl	%esp, %ebp
	.cfi_escape 0x10,0x5,0x2,0x75,0
	pushl	%edi
	pushl	%esi
	pushl	%ebx
	pushl	%ecx
	.cfi_escape 0xf,0x3,0x75,0x70,0x6
	.cfi_escape 0x10,0x3,0x2,0x75,0x74
	.cfi_escape 0x10,0x6,0x2,0x75,0x78
	.cfi_escape 0x10,0x7,0x2,0x75,0x7c
	subl	$88, %esp
	movl	%ecx, %ebx
	movl	%esp, %eax
	movl	%eax, -76(%ebp)
	movl	$16, -60(%ebp)
	movl	$1, -56(%ebp)
	movl	$0, -52(%ebp)
	movl	-60(%ebp), %eax
	leal	-1(%eax), %edx
	movl	%edx, -36(%ebp)
	leal	15(%eax), %edx
	movl	$16, %eax
	subl	$1, %eax
	addl	%edx, %eax
	movl	$16, -80(%ebp)
	movl	$0, %edx
	divl	-80(%ebp)
	imull	$16, %eax, %eax
	subl	%eax, %esp
	leal	16(%esp), %eax
	addl	$15, %eax
	shrl	$4, %eax
	sall	$4, %eax
	movl	%eax, -32(%ebp)
	cmpl	$1, (%ebx)
	jle	.L2
	cmpl	$4, (%ebx)
	jle	.L3
.L2:
	movl	$.LC0, (%esp)
	call	puts
	movl	$.LC1, (%esp)
	call	puts
	movl	$1, (%esp)
	call	exit
.L3:
	movl	$.LC2, -48(%ebp)
	movl	$.LC3, -44(%ebp)
	movl	4(%ebx), %eax
	addl	$4, %eax
	movl	(%eax), %eax
	movl	%eax, %edx
	movl	$.LC4, %eax
	movl	$3, %ecx
	movl	%edx, %esi
	movl	%eax, %edi
	repz cmpsb
	seta	%dl
	setb	%al
	movl	%edx, %ecx
	subb	%al, %cl
	movl	%ecx, %eax
	movsbl	%al, %eax
	testl	%eax, %eax
	jne	.L4
	movl	$.LC5, -48(%ebp)
	movl	$.LC6, %edx
	movl	4(%ebx), %eax
	addl	$8, %eax
	movl	(%eax), %eax
	movl	%edx, 4(%esp)
	movl	%eax, (%esp)
	call	fopen
	movl	%eax, -40(%ebp)
	cmpl	$0, -40(%ebp)
	jne	.L5
	movl	$.LC7, (%esp)
	call	puts
	movl	$.LC1, (%esp)
	call	puts
	movl	$1, (%esp)
	call	exit
.L5:
	movl	$.LC6, %edx
	movl	4(%ebx), %eax
	addl	$8, %eax
	movl	(%eax), %eax
	movl	%edx, 4(%esp)
	movl	%eax, (%esp)
	call	fopen
	movl	%eax, -40(%ebp)
	jmp	.L14
.L4:
	movl	4(%ebx), %eax
	addl	$4, %eax
	movl	(%eax), %eax
	movl	%eax, %edx
	movl	$.LC8, %eax
	movl	$4, %ecx
	movl	%edx, %esi
	movl	%eax, %edi
	repz cmpsb
	seta	%dl
	setb	%al
	movl	%edx, %ecx
	subb	%al, %cl
	movl	%ecx, %eax
	movsbl	%al, %eax
	testl	%eax, %eax
	jne	.L7
	movl	$.LC9, -48(%ebp)
	movl	$.LC10, -44(%ebp)
	movl	$.LC6, %edx
	movl	4(%ebx), %eax
	addl	$8, %eax
	movl	(%eax), %eax
	movl	%edx, 4(%esp)
	movl	%eax, (%esp)
	call	fopen
	movl	%eax, -40(%ebp)
	cmpl	$0, -40(%ebp)
	jne	.L8
	movl	$.LC7, (%esp)
	call	puts
	movl	$.LC1, (%esp)
	call	puts
	movl	$1, (%esp)
	call	exit
.L8:
	movl	$.LC6, %edx
	movl	4(%ebx), %eax
	addl	$8, %eax
	movl	(%eax), %eax
	movl	%edx, 4(%esp)
	movl	%eax, (%esp)
	call	fopen
	movl	%eax, -40(%ebp)
	jmp	.L14
.L7:
	movl	4(%ebx), %eax
	addl	$4, %eax
	movl	(%eax), %eax
	movl	%eax, %edx
	movl	$.LC11, %eax
	movl	$3, %ecx
	movl	%edx, %esi
	movl	%eax, %edi
	repz cmpsb
	seta	%dl
	setb	%al
	movl	%edx, %ecx
	subb	%al, %cl
	movl	%ecx, %eax
	movsbl	%al, %eax
	testl	%eax, %eax
	jne	.L9
	movl	$.LC12, %eax
	movl	$version, 4(%esp)
	movl	%eax, (%esp)
	call	printf
	movl	$1, (%esp)
	call	exit
.L9:
	movl	4(%ebx), %eax
	addl	$4, %eax
	movl	(%eax), %eax
	movl	%eax, %edx
	movl	$.LC13, %eax
	movl	$3, %ecx
	movl	%edx, %esi
	movl	%eax, %edi
	repz cmpsb
	seta	%dl
	setb	%al
	movl	%edx, %ecx
	subb	%al, %cl
	movl	%ecx, %eax
	movsbl	%al, %eax
	testl	%eax, %eax
	jne	.L10
	movl	4(%ebx), %eax
	addl	$8, %eax
	movl	(%eax), %eax
	movl	%eax, (%esp)
	call	atoi
	movl	%eax, -60(%ebp)
	movl	$.LC6, %edx
	movl	4(%ebx), %eax
	addl	$12, %eax
	movl	(%eax), %eax
	movl	%edx, 4(%esp)
	movl	%eax, (%esp)
	call	fopen
	movl	%eax, -40(%ebp)
	cmpl	$0, -40(%ebp)
	jne	.L11
	movl	$.LC7, (%esp)
	call	puts
	movl	$.LC1, (%esp)
	call	puts
	movl	$1, (%esp)
	call	exit
.L11:
	movl	$.LC6, %edx
	movl	4(%ebx), %eax
	addl	$12, %eax
	movl	(%eax), %eax
	movl	%edx, 4(%esp)
	movl	%eax, (%esp)
	call	fopen
	movl	%eax, -40(%ebp)
	jmp	.L14
.L10:
	movl	4(%ebx), %eax
	addl	$4, %eax
	movl	(%eax), %eax
	movl	%eax, %edx
	movl	$.LC14, %eax
	movl	$3, %ecx
	movl	%edx, %esi
	movl	%eax, %edi
	repz cmpsb
	seta	%dl
	setb	%al
	movl	%edx, %ecx
	subb	%al, %cl
	movl	%ecx, %eax
	movsbl	%al, %eax
	testl	%eax, %eax
	jne	.L12
	movl	$.LC15, (%esp)
	call	system
	movl	%eax, -28(%ebp)
	cmpl	$-1, -28(%ebp)
	jne	.L13
	movl	$.LC16, (%esp)
	call	puts
.L13:
	movl	$1, (%esp)
	call	exit
.L12:
	movl	$.LC6, %edx
	movl	4(%ebx), %eax
	addl	$4, %eax
	movl	(%eax), %eax
	movl	%edx, 4(%esp)
	movl	%eax, (%esp)
	call	fopen
	movl	%eax, -40(%ebp)
	cmpl	$0, -40(%ebp)
	jne	.L14
	movl	$.LC7, (%esp)
	call	puts
	movl	$.LC1, (%esp)
	call	puts
	movl	$1, (%esp)
	call	exit
.L22:
	movl	-60(%ebp), %edx
	movl	-32(%ebp), %eax
	movl	-40(%ebp), %ecx
	movl	%ecx, 12(%esp)
	movl	%edx, 8(%esp)
	movl	$1, 4(%esp)
	movl	%eax, (%esp)
	call	fread
	movl	%eax, -56(%ebp)
	movl	-44(%ebp), %eax
	movl	-52(%ebp), %edx
	movl	%edx, 4(%esp)
	movl	%eax, (%esp)
	call	printf
	movl	$0, -64(%ebp)
	jmp	.L15
.L16:
	movl	-64(%ebp), %eax
	leal	1(%eax), %edx
	movl	-32(%ebp), %eax
	movzbl	(%eax,%edx), %eax
	movzbl	%al, %ecx
	movl	-32(%ebp), %edx
	movl	-64(%ebp), %eax
	addl	%edx, %eax
	movzbl	(%eax), %eax
	movzbl	%al, %edx
	movl	-48(%ebp), %eax
	movl	%ecx, 8(%esp)
	movl	%edx, 4(%esp)
	movl	%eax, (%esp)
	call	printf
	addl	$2, -64(%ebp)
.L15:
	movl	-64(%ebp), %eax
	cmpl	-60(%ebp), %eax
	jl	.L16
	movl	$9, (%esp)
	call	putchar
	movl	$0, -64(%ebp)
	jmp	.L17
.L21:
	movl	-32(%ebp), %edx
	movl	-64(%ebp), %eax
	addl	%edx, %eax
	movzbl	(%eax), %eax
	cmpb	$31, %al
	jbe	.L18
	movl	-32(%ebp), %edx
	movl	-64(%ebp), %eax
	addl	%edx, %eax
	movzbl	(%eax), %eax
	testb	%al, %al
	jns	.L19
.L18:
	movl	$46, (%esp)
	call	putchar
	jmp	.L20
.L19:
	movl	-32(%ebp), %edx
	movl	-64(%ebp), %eax
	addl	%edx, %eax
	movzbl	(%eax), %eax
	movsbl	%al, %eax
	movl	%eax, (%esp)
	call	putchar
.L20:
	addl	$1, -64(%ebp)
.L17:
	movl	-64(%ebp), %eax
	cmpl	-60(%ebp), %eax
	jl	.L21
	movl	-60(%ebp), %eax
	addl	%eax, -52(%ebp)
.L14:
	cmpl	$0, -56(%ebp)
	jne	.L22
	movl	$.LC10, (%esp)
	call	puts
	movl	-40(%ebp), %eax
	movl	%eax, (%esp)
	call	fclose
	movl	$0, %eax
	movl	-76(%ebp), %esp
	leal	-16(%ebp), %esp
	popl	%ecx
	.cfi_def_cfa 1, 0
	.cfi_restore 1
	popl	%ebx
	.cfi_restore 3
	popl	%esi
	.cfi_restore 6
	popl	%edi
	.cfi_restore 7
	popl	%ebp
	.cfi_restore 5
	leal	-4(%ecx), %esp
	.cfi_def_cfa 4, 4
	ret
	.cfi_endproc
.LFE0:
	.size	main, .-main
	.ident	"GCC: (Ubuntu/Linaro 4.6.3-1ubuntu5) 4.6.3"
	.section	.note.GNU-stack,"",@progbits


print.h
#ifndef PRINT_H
#define PRINT_H

/**********************************************************
 * print_usage: Prints help message.                       *
 **********************************************************/

void print_usage(void);

/**********************************************************
 * print_version: Prints Version.                         *
 **********************************************************/

void print_version(void);

/**********************************************************
 * print_hex_upper: Prints upper case hex letters.        *
 **********************************************************/

int print_hex_upper(void);
 
/**********************************************************
 * print_col_width: Prints based on collum width.         *
 **********************************************************/

int print_col_width(int width);




#endif


hfd.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>


char  version[40] = "hfd v001.6 24oct12 by Ed Rantanen";

#define PRINT(print_method,elemnt1,elemnt2)  printf(print_method, n[incr], n[incr+1])
#define PRINT_Num_Elements(l_fmt, elementGroup) printf(l_fmt, elementGroup)

void print_k() {
      printf("printing k from case switch\n");
    }

void print_default() {
      printf("printing default from case switch\n");
    }




/*****************************************
 *   hfd -> hex file dump
 *
 * This program reads in data, and does
 * a display depending on conditionals 
 * requested. 
 *
 * by: Ed Rantanen
 * updated: 15 Nov 2012
 *
 * updates and documentation 
 *      at github.com/erantanen/hfd
 *
 ****************************************/

int  main(int argc, char *argv[])
{
/* initial declarations */
  int option = 0;
  int incr;
  int dataBits = 16;
  int numberofElements = 1;
  int elementGroup = 0;       /*number of bits read */
  unsigned char  n[dataBits]; /* holder of data */
/*  char  *file_in;*/
  char  *print_method;        /* print fmt for block */
  char  *l_fmt;               /* print fmt for lines */
  FILE  *filehandle_in;


/*************************
 * parsing argv elements *
 *************************/

   while((option=getopt(argc, argv, "k")) != -1) {
            switch (option) {
                  case 'k' : print_k();
                      break;
                    default: print_default();
                      exit(0);
             } //swtich
    } //while



   if(argc < 2 || argc > 4) {
      printf("\n 1 In correct usage,\n \' hfd -help \'\n");
      printf("Try for a better understanding of how to use hfd.\n\n");
      exit(1);
    }

   /* prep for switches */
       print_method = " %02x%02x ";  /* default print */
       l_fmt = "\n %06x:";          /* default line numbering */
    
        if(strcmp(argv[1], "-u") == 0) {
              
            print_method = " %02X%02X ";         
               if((filehandle_in = fopen( argv[2], "r")) == NULL) {
                 printf("\n 2 In correct usage,\n \' hfd -help \'\n");
                 printf("Try for a better understanding of how to use hfd.\n\n");
                 exit(1);
                } else {
                     filehandle_in = fopen( argv[2], "r");
                }

           } else if(strcmp(argv[1], "-I") == 0) {

                   int c=0;
                   printf("This is for command line input,\n enter data when finished hit return\n");
                   while((c=getchar()) != 10){
                     printf("%x ",c);
                    }

                   printf("\nThis is the end of command line input\n");
                   exit(0); 
           } else if(strcmp(argv[1], "-d") == 0) {

             print_method = " %3d %3d";

               if((filehandle_in = fopen( argv[2], "r")) == NULL) {
                 printf("\n 2 In correct usage,\n \' hfd -help \'\n");
                 printf("Try for a better understanding of how to use hfd.\n\n");
                 exit(1);
                } else {
                     filehandle_in = fopen( argv[2], "r");
                }
         } else if(strcmp(argv[1], "-ps") == 0) {
                         
            print_method = "%02x%02x";
            l_fmt = "\n";               
               if((filehandle_in = fopen( argv[2], "r")) == NULL) {
                 printf("\n 2 In correct usage,\n \' hfd -help \'\n");
                 printf("Try for a better understanding of how to use hfd.\n\n");
                 exit(1);
                } else {
                     filehandle_in = fopen( argv[2], "r");
                }

        } else if(strcmp(argv[1], "-v") == 0) {
            printf("\n%s \n\n", version);
            exit(1);

        } else if(strcmp(argv[1], "-c") == 0) {
        
                dataBits = atoi(argv[2]); /* convert arg to int */
              if((filehandle_in = fopen( argv[3], "r")) == NULL) {
                 printf("\n 2 In correct usage,\n \' hfd -help \'\n");
                 printf("Try for a better understanding of how to use hfd.\n\n");
                 exit(1);
                } else {
                     filehandle_in = fopen( argv[3], "r");
                }   

        } else if(strcmp(argv[1], "-h") == 0) {
           int systemresult =  system("man ./hfd_man");
           if(systemresult == -1) {
              printf("Unable to locate help file\n");
             }
            exit(1);

        } else if((filehandle_in = fopen( argv[1], "r")) == NULL) {
            printf("\n 2 In correct usage,\n \' hfd -help \'\n");
            printf("Try for a better understanding of how to use hfd.\n\n");
            exit(1);
        }

 /* begin file process */

 while( numberofElements != 0) {

  numberofElements = fread(n, sizeof(char), dataBits, filehandle_in);
    
 /* begin internal print block */
     
     PRINT_Num_Elements(l_fmt, elementGroup); /* line numbering */
   
     for(incr = 0; incr < dataBits ; incr+=2){  PRINT(print_method, n[incr], n[incr+1]); } 

     printf("\t");

     for(incr = 0; incr < dataBits ; incr++){

     /*  usable ascii symbols are 33 to 126 */
      if ( n[incr] < 32 || n[incr] > 127 ){
          printf(".");
        }
      else 
       {
         printf("%c", (char)n[incr]);
       }
    }

   elementGroup = elementGroup + dataBits;
 }

/* end of while */

    printf("\n\n"); /* simple clean up after run */

fclose(filehandle_in);

return(0); 
}
